src/js/components/DataTableColumns/DataTableColumns.js

```diff
@@ -0,0 +1,163 @@
+import { CheckBoxGroup } from '../CheckBoxGroup';
+import { DataForm, formColumnsKey } from '../Data';
+import { FormContext } from '../Form/FormContext';
+import { DropButton } from '../DropButton';
+import { List } from '../List';
+import { Tab } from '../Tab';
+import { Tabs } from '../Tabs';
+import { TextInput } from '../TextInput';
+import { DataContext } from '../../contexts/DataContext';
+import { MessageContext } from '../../contexts/MessageContext';
+import { DataTableColumnsPropTypes } from './propTypes';
+
+const dropProps = {
+  align: { top: 'bottom', left: 'left' },
+};
+
+const tabsProps = {
+  drop: { pad: 'small' },
+  noDrop: { justify: 'start' },
+};
+
+// options can either be an array of property names or an array of objects.
+// The form value always uses an array of property names.
+const optionsToValue = (options) =>
+  options.map((o) => (typeof o === 'object' && o.property) || o) || [];
+
+// Content is a separate component since it might be getting its form context
+// from the DataForm rendered inside DataTableColumns.
+const Content = ({ drop, options, ...rest }) => {
+  const { id: dataId, messages } = useContext(DataContext);
+  const { useFormInput } = useContext(FormContext);
+  const { format } = useContext(MessageContext);
+
+  // In cases the user searches for a particular option, render
+  // the filtered list of options.
+  const [search, setSearch] = useState('');
+  const [filteredOptions, setFilteredOptions] = useState(options);
+
+  // Note whether options are objects so we can set up the *key properties
+  // as needed when rendering.
+  const objectOptions = useMemo(
+    () => options && options.length && typeof options[0] === 'object',
+    [options],
+  );
+
+  // 'value' is an array of property names
+  const [value, setValue] = useFormInput({
+    name: formColumnsKey,
+    initialValue: optionsToValue(options?.slice(0, 3)),
+  });
+
+  // When the user searches, updated the filtered options based on the
+  // search string.
+  const onSearch = useCallback(
+    (nextSearch) => {
+      let nextFilteredOptions = options;
+      if (nextSearch) {
+        const lowerSearch = nextSearch.toLowerCase();
+        nextFilteredOptions = options.filter(
+          (o) =>
+            (!objectOptions && o.toLowerCase().includes(lowerSearch)) ||
+            (objectOptions &&
+              (o.property.toLowerCase().includes(lowerSearch) ||
+                o.label.toLowerCase().includes(lowerSearch))),
+        );
+      }
+      setSearch(nextSearch);
+      setFilteredOptions(nextFilteredOptions);
+    },
+    [objectOptions, options],
+  );
+
+  return (
+    <Box>
+      <Tabs {...tabsProps[drop ? 'drop' : 'noDrop']} {...rest}>
+        <Tab
+          id={`${dataId}--select-columns-tab`}
+          title={format({
+            id: 'dataTableColumns.select',
+            messages: messages?.dataTableColumns,
+          })}
+        >
+          <Box pad={{ top: 'small' }}>
+            <TextInput
```

Comment:
I am unable to type certain characters into the search. For example on the inline story I can't type 'p' into the search